client/src/sim.rs

```diff
@@ -165,6 +166,29 @@ impl Sim {
         self.selected_material
     }
 
+    /// Return the list of EntityIds corresponding to all inventory items matching the given material
+    pub fn inventory_contents_matching_material(&self, material: Material) -> Vec<EntityId> {
```

Comment:
I tried to do that at first but didn't think it was possible, so I likely need some Rust assistance in figuring out how to do this. Even if I simplify this function to a one-liner that's more likely to panic:
```rust
pub fn inventory_contents_matching_material(
    &self,
    material: Material,
) -> impl Iterator<Item = EntityId> + '_ {
    self.world
        .get::<&Inventory>(self.local_character.unwrap())
        .unwrap()
        .contents
        .iter()
        .copied()
        .filter(|e| {
            let Some(entity) = self.entity_ids.get(e) else {
                return false;
            };
            self.world
                .get::<&Material>(*entity)
                .is_ok_and(|m| *m == material)
        })
}
```
I get the following compiler diagnostic
```
error[E0515]: cannot return value referencing temporary value
   --> client\src\sim.rs:174:9
    |
174 | //         self.world
175 | ||             .get::<&Inventory>(self.local_character.unwrap())
176 | ||             .unwrap()
    | ||_____________________- temporary value created here
177 | |              .contents
...   |
186 | |                      .is_ok_and(|m| *m == material)
187 | |              })
    | |_______________^ returns a value referencing data owned by the current function
    |
    = help: use `.collect()` to allocate the iterator
```

It seems like it should be theoretically possible to handle this correctly (hopefully without using a custom iterator), since the temporary `hec::entity_ref::Ref<Inventory>` that gets created could live in the iterator that gets returned instead of the stack, but I haven't figured out how to get that to happen.

Reply:
I was able to get this to work with a custom iterator:
```rust
impl Inventory {
    pub fn iter(inventory: hecs::Ref<Self>) -> InventoryIterator<'_> {
        InventoryIterator {
            inventory,
            index: 0,
        }
    }
}

pub struct InventoryIterator<'a> {
    inventory: hecs::Ref<'a, Inventory>,
    index: usize,
}

impl Iterator for InventoryIterator<'_> {
    type Item = EntityId;

    fn next(&mut self) -> Option<Self::Item> {
        let result = self.inventory.contents.get(self.index).copied();
        self.index += 1;
        result
    }
}
```
used in the following way
```rust
let local_character = self.local_character.unwrap();
self.world
    .get::<&Inventory>(local_character)
    .into_iter()
    .flat_map(move |inventory| {
        Inventory::iter(inventory).filter(move |e| {
            let Some(entity) = self.entity_ids.get(e) else {
                return false;
            };
            self.world
                .get::<&Material>(*entity)
                .is_ok_and(|m| *m == material)
        })
    })
```

In the back of my mind, I've been assuming that `hecs::Ref` was `Copy`, even though it clearly couldn't be. It's possible that I can add some `clone()` calls to get this to work without needing to write a custom iterator. I just need to be careful not to accidentally `clone()` the `Inventory` beneath. Perhaps I should just be explicit instead of using dot syntax for that.

Reply:
Actually, I think getting it to work without a custom iterator is impossible after all. As far as I know, all the default iterators for vectors are structs that contain a pointer to a slice. However, we need to store the `Ref<Inventory>` itself, which effectively falls under the pattern of "storing a thing and a reference to that things in the same struct".

Although `hecs::Ref` behaves like a reference, so the thing it's borrowing doesn't actually move, I don't think I can get Rust to recognize that without using `unsafe` or abusing unsoundness.

For now, I'll go ahead with the `FnOnce` approach, possibly using helper methods to make its use clearer. I don't like the idea of having to create a custom iterator that has to do extra work per step.

Reply:
Oops, yes, `Ref` necessarily makes this awkward. [`std::iter::from_fn`](https://doc.rust-lang.org/std/iter/fn.from_fn.html) can be a good way to bodge a quick custom iterator, but the result is still less readable, so it's not necessarily worth pursuing.

Reply:
Yeah, that's unfortunate, since it definitely seems like something that should be able to work somewhat conveniently, but for now, I'll stick with the two entirely separate functions.